packages/punchcard/lib/core/code.ts

```diff
@@ -98,43 +107,56 @@ export namespace Code {
         externals,
         plugins
       };
-      const compiler: Compiler = webpack(config);
-
-      fs.writeFileSync(path.join(codePath, 'index.js'), `
-  require('./app');
-  const entrypointId = process.env.entrypoint_id;
-  if (!entrypointId) {
-    throw new Error('entrypoint_id environment variable is missing');
-  }
-  const state = global[Symbol.for('punchcard.global')];
-  if (!state) {
-    throw new Error('global state missing');
-  }
-  const entrypoint = state.entrypoints[entrypointId];
-  if (!entrypoint) {
-    throw new Error('entrypoint with id "' + entrypointId + '" does not exist');
-  }
-  var handler;
-  exports.handler = async (event, context) => {
-    if (!handler) {
-      handler = await entrypoint.entrypoint[Symbol.for('Runtime.get')]();
-    }
-    return await handler(event, context);
-  }
-  `);
-
-      // TODO: fix this - it's async so the code asset is usually built before the file is written
-      compiler.run((err: any, stats: any) => {
-        if (err) {
-          console.error(err);
-        } else if (stats?.compilation?.errors?.length) {
-          // Sometimes the err will be null, but the errors array won't, so map each
-          // one individually to the console
-          stats.compilation.errors.map((s: any) => console.error(s.message ?? s));
-        }
-        (app as any)[symbol] = lambda.Code.asset(codePath);
-        cb((app as any)[symbol]);
+      const compiler = webpack(config);
+
+      await fs.writeFile(path.join(codePath, 'index.js'), indexFile);
+
+      const asset = await new Promise((resolve, reject) => {
+        // this must be called before the CDK Construct tree is built because it is async and Construct constructors
+        // like Assets require synchronous instantiation
+        compiler.run((err: any, stats: any) => {
+          if (err) {
+            reject(err);
+          } else if (stats?.compilation?.errors?.length) {
+            // Sometimes the err will be null, but the errors array won't, so map each
+            // one individually to the console
+            stats.compilation.errors.map((s: any) => console.error(s.message ?? s));
+          }
+          resolve(lambda.Code.asset(codePath));
+        });
       });
+
+      // cache the asset
+      (app as any)[symbol] = asset;
     }
+
+    // return the cached one
+    return (app as any)[symbol];
   }
 }
+
+// TODO: put in a static file?
```

Comment:
would be the best bet, a simple `fs.readSync().split('{{{placeholder}}}').join(value)` would do as a poor's man templating engine haha